receiver/elasticsearchreceiver/client.go

```diff
@@ -0,0 +1,153 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package elasticsearchreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/elasticsearchreceiver"
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/elasticsearchreceiver/internal/model"
+)
+
+var (
+	errUnauthenticated = errors.New("status 401, unauthenticated")
+	errUnauthorized    = errors.New("status 403, unauthorized")
+)
+
+// elasticsearchClient defines the interface to retrieve metrics from an Elasticsearch cluster.
+type elasticsearchClient interface {
+	NodeStats(ctx context.Context, nodes []string) (*model.NodeStats, error)
+	ClusterHealth(ctx context.Context) (*model.ClusterHealth, error)
+}
+
+// defaultElasticsearchClient is the main implementation of elasticsearchClient.
+// It retrieves the required metrics from Elasticsearch's REST api.
+type defaultElasticsearchClient struct {
+	client     *http.Client
+	endpoint   *url.URL
+	authHeader string
+	logger     *zap.Logger
+}
+
+var _ elasticsearchClient = (*defaultElasticsearchClient)(nil)
+
+func newElasticsearchClient(logger *zap.Logger, client *http.Client, endpoint *url.URL, username, password string) *defaultElasticsearchClient {
```

Comment:
Would it make more sense to just pass the config, use it to create the client here? Seems unusual that we would expect it to be created elsewhere first.